src/Catalyst.KBucket/KBucket.cs

```diff
@@ -0,0 +1,503 @@
+        /// <param name="contact">
+        ///   When this method returns, contains the <see cref="IContact"/> associated 
+        ///   with the <paramref name="id"/>, if the key is found; otherwise, <b>null</b>.
+        ///   This parameter is passed uninitialized.
+        /// </param>
+        /// <returns>
+        ///  <b>true</b> if the <paramref name="id"/> is found; otherwise <b>false</b>.
+        /// </returns>
+        public bool TryGet(byte[] id, out T contact)
+        {
+            _rwlock.EnterReadLock();
+            try
+            {
+                contact = _Get(id);
+                return contact != null;
+            }
+            finally
+            {
+                _rwlock.ExitReadLock();
+            }
+        }
+
+        /// <inheritdoc />
+        public void CopyTo(T[] array, int arrayIndex)
+        {
+            foreach (var contact in this) array[arrayIndex++] = contact;
+        }
+
+        /// <inheritdoc />
+        public IEnumerator<T> GetEnumerator()
+        {
+            _rwlock.EnterReadLock();
+            try
+            {
+                foreach (var contact in Root.AllContacts()) yield return contact;
+            }
+            finally
+            {
+                _rwlock.ExitReadLock();
+            }
+        }
+
+        /// <inheritdoc />
+        public bool Remove(T item)
+        {
+            Validate(item);
+
+            _rwlock.EnterWriteLock();
+            try
+            {
+                return _Remove(item.Id);
+            }
+            finally
+            {
+                _rwlock.ExitWriteLock();
+            }
+        }
+
+        /// <inheritdoc />
+        IEnumerator IEnumerable.GetEnumerator() { return GetEnumerator(); }
+
+        /// <summary>
+        ///   Check that contact is correct.
+        /// </summary>
+        /// <param name="contact">
+        ///   The <see cref="IContact"/> to validate.
+        /// </param>
+        /// <exception cref="ArgumentNullException">
+        ///   When <paramref name="contact"/> is null or its <see cref="IContact.Id"/>
+        ///   is null or empty.
+        /// </exception>
+        private void Validate(T contact)
+        {
+            if (contact == null) throw new ArgumentNullException("contact");
+
+            if (contact.Id == null || contact.Id.Length == 0) throw new ArgumentNullException("contact.Id");
+        }
+
+        /// <summary>
+        ///   Add the contact.
+        /// </summary>
+        /// <param name="contact"></param>
+        /// <param name="ping"></param>
+        /// <returns>
+        ///   <b>true</b> if the <paramref name="contact"/> was added; otherwise, 
+        ///   <b>false</b> and a <see cref="Ping"/> event should be raised.
+        /// </returns>
+        private bool _Add(T contact, out PingEventArgs<T> ping)
+        {
+            ping = null;
+
+            var bitIndex = 0;
+            var node = Root;
+
+            while (node.Contacts == null)
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Add curly braces around the nested statement(s) in this 'while' block.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)